ocrd/ocrd/resolver.py

```diff
@@ -94,55 +113,40 @@ def workspace_from_url(self, mets_url, dst_dir=None, clobber_mets=False, mets_ba
             dst_dir (string, None): Target directory for the workspace
             clobber_mets (boolean, False): Whether to overwrite existing mets.xml. By default existing mets.xml will raise an exception.
             download (boolean, False): Whether to download all the files
-            baseurl (string, None): Base URL for resolving relative file locations
+            src_baseurl (string, None): Base URL for resolving relative file locations
 
         Returns:
             Workspace
         """
-        if dst_dir and not dst_dir.startswith('/'):
-            dst_dir = abspath(dst_dir)
 
         if mets_url is None:
-            if baseurl is None:
-                raise Exception("Must pass mets_url and/or baseurl to workspace_from_url")
-            else:
-                mets_url = 'file://%s/%s' % (baseurl, mets_basename if mets_basename else 'mets.xml')
-        if baseurl is None:
-            baseurl = mets_url.rsplit('/', 1)[0]
-        log.debug("workspace_from_url\nmets_url='%s'\nbaseurl='%s'\ndst_dir='%s'", mets_url, baseurl, dst_dir)
-
-        # resolve to absolute
-        if '://' not in mets_url:
-            mets_url = 'file://%s' % abspath(mets_url)
-
-        if dst_dir is None:
-            # if mets_url is a file-url assume working directory is source directory
-            if mets_url.startswith('file://'):
-                # if dst_dir was not given and mets_url is a file assume that
-                # dst_dir should be the directory where the mets.xml resides
-                dst_dir = dirname(mets_url[len('file://'):])
-            else:
-                dst_dir = tempfile.mkdtemp(prefix=TMP_PREFIX)
-                log.debug("Creating workspace '%s' for METS @ <%s>", dst_dir, mets_url)
+            raise ValueError("Must pass 'mets_url' workspace_from_url")
 
         # if mets_basename is not given, use the last URL segment of the mets_url
         if mets_basename is None:
-            mets_basename = mets_url \
-                .rsplit('/', 1)[-1] \
-                .split('?')[0] \
-                .split('#')[0]
-
-        dst_mets = join(dst_dir, mets_basename)
-        log.debug("Copying mets url '%s' to '%s'", mets_url, dst_mets)
-        if 'file://' + dst_mets == mets_url:
-            log.debug("Target and source mets are identical")
-        else:
-            if exists(dst_mets) and not clobber_mets:
-                raise Exception("File '%s' already exists but clobber_mets is false" % dst_mets)
+            mets_basename = nth_url_segment(mets_url, -1)
+
+        # If src_baseurl wasn't given, determine from mets_url by removing last url segment
+        if not src_baseurl:
+            last_segment = nth_url_segment(mets_url)
+            src_baseurl = remove_non_path_from_url(remove_non_path_from_url(mets_url)[:-len(last_segment)])
```

Comment:
This looks like implicit magic. Do we need it?

Reply:
You may ignore this question if it is dumb.

Reply:
It's not a dumb question, code is a bit convoluted but necessarily so.

The problem is that URLs can contain fragment identifiers and query strings and trailing slashes. The algorithm here ensures that `src_baseurl` of

`http://foo/path/to/mets.xml?yay-query-string#frag`

resolves to

`http://foo/path/to`

without trailing slash, query string or fragment identifier.

`